# First attempt at a Jupyter Notebook

This is just a first attempt at exploring how one of these notebooks works. Here's a recursive fibonacci number algorithm.


In [9]:
def fib(n):
    if n <= 2:
        # Base case the 1st and 2nd fibonnaci numbers are 1 by definition
        return 1
    
    else:
        # Recursive case
        return fib(n-2) + fib(n-1)

print(fib(10))

55


This algorithm is simple to read, but it is brutally inefficient.

A fibonacci function should be really fast because it should only have to do _n_ - 1 additions. But, this function is actually going to do _n_*_n_-1 additions. This happens because the recursion is _dumb_. It recalculates values it has already seen before.

For example, the first time the function is called, n = 10. So the function reaches the recursive case and branches off into two separate function calls: `fib(8)` and `fib(9)`.

The problem here is that `fib(8)` and `fib(9)` don't know about each other. Both will each branch on their own into two new function calls. And those function calls will repeat work. For example, both will need to call `fib(7)`. That repeated work means the computer is doing more work than it needs to. 

It should only need to call `fib()` 10 total times. So let's see how many times it is calling `fib()`

In [4]:
counter = 0
def fib(n):
    global counter
    counter += 1
    if n <= 2:
        return 1
    
    else:
        return fib(n-2) + fib(n-1)

fib(10)
print(f'Fib() was called {counter} times.')

Fib() was called 109 times.


Holy moly! That's too many times! Let's track what _n_s are passed into `fib(n)`


In [3]:
n_counter = {}

def fib(n):
    # Update our our counts
    global n_counter
    if not n in n_counter:
        n_counter[n] = 1
    else: 
        n_counter[n] += 1
    
    if n <= 2:
        return 1
    
    else:
        return fib(n-2) + fib(n-1)

fib(10)

# Print out how many times each was called
for n in n_counter:
    print(f'f({n}) called {n_counter[n]} times')

f(10) called 1 times
f(8) called 2 times
f(6) called 5 times
f(4) called 13 times
f(2) called 34 times
f(3) called 21 times
f(1) called 21 times
f(5) called 8 times
f(7) called 3 times
f(9) called 1 times


The higher *n* values aren't too bad, but `f(2)` is called 34 times! What!?

Well, that's because as our recursive function is called again and again, new branches keep forking off. Each branch is independent of the others and can't reuse the answers from the other branches. So, every branch has to compute `f(2)`. On it's own. And, sure, it can do that pretty quickly, but at some point the number of branches gets **SO** big that it doesn't matter how quickly the computer can handle any step, there are just too many steps.

This time, let's try `fib(20)` and see how many times just `fib(2)` is called.

In [8]:
fib_of_anything_counter = 0
fib_of_two_counter = 0

def fib(n):
    # Update our our counts
    global fib_of_two_counter
    global fib_of_anything_counter
    
    fib_of_anything_counter += 1
    if n == 2:
        fib_of_two_counter += 1
    
    if n <= 2:
        return 1
    else:
        return fib(n-2) + fib(n-1)

fib(20)

print(f'fib(2) was called {fib_of_two_counter} times.')
print(f'fib(anything) was called {fib_of_anything_counter} times.')

fib(2) was called 4181 times.
fib(anything) was called 13529 times.


Woah! `fib(2)` was called 4181 times and `fib()` was called a total of 13529 times.

In Computer Science, we measure the speed of an algorithm using Big O Notation which is meant to capture how the runtime grows as the number of elements increases. For `fib(10)` the algorithm ran 109 times. You can think of it as having 109 steps. But, when we doubled *n* the number of steps shot up to 13,529! That's about a 124 times increase in the number of steps. Big O Notation is just a way of describing how the number of steps changed as we increase *n*.

If the number of steps hadn't changed, this would have been a constant runtime, abbreviated as O(1). If increasing n by 10 increased the steps by 10, the runtime would be linear, O(n). This is worse than both of those. It's something like O(n^2).

But, we can do better. Here's a non recursive apprach.

In [14]:
def fib(n):
    nth = n

    if n <= 2:
        return 1
    
    previous_previous_fib = 0
    previous_fib = 1
    current_fib = 1

    counter = 0

    while n > 2:
        counter += 1

        previous_previous_fib = previous_fib
        previous_fib = current_fib
        current_fib = previous_previous_fib + previous_fib
        
        n -= 1
    
    print(f'fib({nth}) used {counter} steps to solve this.')
    return current_fib

fib(10)
fib(20)

fib(10) used 8 steps to solve this.
fib(20) used 18 steps to solve this.


6765

How 'bout them apples. This time `fib(20)` Only took 18 steps. That's a hair better than 13,529 steps.

Notice, also, that increase *n* by 10 only increase our steps by 10. That means this algorithm is linear! O(n).